### PR TITLE
Refactoring the scanner service to offer a class for performin one si…

### PR DESCRIPTION
**Pull request by arcturus**
…ngle check on location.

### CODE REVIEW

**Comment by arcturus** [comment]
@wilsonpage r?

This is a step closer to reuse the new OneShotLocation class in region detection.

**Comment by arcturus** [comment]
@wilsonpage this needs ❤️ so if you feel jetlagged remember this is waiting for you.

File: MagnetScannerIOS/OneShotLocation.swift [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+        
+        guard location.horizontalAccuracy <= self.locationManager.desiredAccuracy else {
+            NSLog("Discarding location \(location) because accuracy \(location.horizontalAccuracy) is \(self.locationManager.desiredAccuracy)")
+            return;
+        }
+        
+        stop()
+        
+        self.callback!(location)
+
```
Comment (wilsonpage) [inline MagnetScannerIOS/OneShotLocation.swift]: NIT: Unnecessary line break

File: MagnetScannerIOS/OneShotLocation.swift [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        
+        guard location.horizontalAccuracy >= 0 else {
+            return
+        }
+        
+        guard bestEffort == nil || bestEffort!.horizontalAccuracy > location.horizontalAccuracy else {
+            return
+        }
+        
+        bestEffort = location
```
Comment (wilsonpage) [inline MagnetScannerIOS/OneShotLocation.swift]: Does `bestEffort` mean: 'last good location'?

File: MagnetScannerIOS/OneShotLocation.swift [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        
+        guard location.horizontalAccuracy <= self.locationManager.desiredAccuracy else {
```
Comment (wilsonpage) [inline MagnetScannerIOS/OneShotLocation.swift]: If accuracy means the same thing in android as ios, magnet-scanner-ios's `100` meters is a lot less accurate than magnet-scanner-android's `20`. Perhaps we can relax magnet-scanner-android?

File: MagnetScannerIOS/ScannerGeolocation.swift [hunk 4]
```diff
@@ -10,88 +10,36 @@ import CoreLocation
 import SwiftyJSON
 
 class ScannerGeolocation: NSObject, CLLocationManagerDelegate, Scanner {
-  let locationManager: CLLocationManager = CLLocationManager()
+  var locationResolver: OneShotLocation = OneShotLocation()
```
Comment (wilsonpage) [inline MagnetScannerIOS/ScannerGeolocation.swift]: I find the naming `OneShotLocation` quite confusing, what is it implying? From a consumer standpoint, `getLocation` seems to be clearer naming to me. Unless I've missed something :)

File: MagnetScannerIOS/OneShotLocation.swift (hunk 2 above)
Comment (arcturus) [reply MagnetScannerIOS/OneShotLocation.swift]: yes, means best location I could get based on the constraints you gave me.

File: MagnetScannerIOS/ScannerGeolocation.swift (hunk 4 above)
Comment (arcturus) [reply MagnetScannerIOS/ScannerGeolocation.swift]: The way of getting location in ios means you will receive plenty of callbacks when you ask for a location.
OneShotLocation means this class will give you one single location, one call, one callback :)

But I'm open to rename anything :P, suggestions welcome.

File: MagnetScannerIOS/NetworkResolver.swift [hunk 5]
```diff
@@ -15,11 +15,11 @@ import SwiftyJSON
 //
 public class NetworkResolver {
   static let API_END_POINT: String = "https://tengam.org/content/v1/search/beacons/"
-  static let RADIUS: Int = 50
+  static let RADIUS: Int = 100
 
   static func resolveLocation(lat: Double, lon: Double, callback: (Array<JSON> -> Void)!) {
```
Comment (wilsonpage) [inline MagnetScannerIOS/NetworkResolver.swift]: NIT Naming: We're not actually resolving the location, we're resolving the item from the location. Something like `fetchItem` or `resolveItem` with `lat` and `lon` as arguments is clearer to me.

**Comment by wilsonpage** [comment]
To summarise this patch:
- Moves all location retrieval logic into a separate class.
- Increases scan virtual beacon scan radius.
- Let me know if I missed anything else :)

Comments:
- Just some naming that could be made more semantic and conventional for readability.
- Otherwise, nice tidy up.

**Comment by wilsonpage** [comment]
Shhiiiiii 🚢 !

File: MagnetScannerIOS/OneShotLocation.swift (hunk 3 above)
Comment (arcturus) [reply MagnetScannerIOS/OneShotLocation.swift]: I like how the android one works tbh.

We could try with a more accurate in ios, but as well, so far we are not being super aggressive with the battery.

File: MagnetScannerIOS/NetworkResolver.swift (hunk 5 above)
Comment (arcturus) [reply MagnetScannerIOS/NetworkResolver.swift]: Well, actually it does ask if there are items in this location, it doesn't resolve items. I think resolveLocation is a good name, we could go for findItems if you want.

BTW, not part of this patch :P

**Comment by arcturus** [comment]
> To summarise this patch:
> 
> ```
> Moves all location retrieval logic into a separate class.
> Increases scan virtual beacon scan radius.
> Let me know if I missed anything else :)
> ```
> 
> Comments:
> 
> ```
> Just some naming that could be made more semantic and conventional for readability.
> Otherwise, nice tidy up.
> ```

You got it crystal clear. The reason for separating the location logic as well is for reusing it outside of the scanner.

File: MagnetScannerIOS/OneShotLocation.swift (hunk 2 above)
Comment (wilsonpage) [reply MagnetScannerIOS/OneShotLocation.swift]: `lastLocation` or `lastValidLocation`

File: MagnetScannerIOS/ScannerGeolocation.swift (hunk 4 above)
Comment (wilsonpage) [reply MagnetScannerIOS/ScannerGeolocation.swift]: `getLocation` work for me.

Comment (arcturus) [reply MagnetScannerIOS/ScannerGeolocation.swift]: It's a class, getLocation does not work.

I can go for LocationHelper.

File: MagnetScannerIOS/NetworkResolver.swift (hunk 5 above)
Comment (wilsonpage) [reply MagnetScannerIOS/NetworkResolver.swift]: BIKESHED: `resolveLocation` suggests the eventual return value will be a location. Much like Node's `url.resolve`. This function is `getItemsFromLocation`, which is different.

File: MagnetScannerIOS/ScannerGeolocation.swift (hunk 4 above)
Comment (wilsonpage) [reply MagnetScannerIOS/ScannerGeolocation.swift]: Sounds good :)

File: MagnetScannerIOS/NetworkResolver.swift (hunk 5 above)
Comment (arcturus) [reply MagnetScannerIOS/NetworkResolver.swift]: Ok.

**Pull request merged**